futures-util/src/io/mod.rs

```diff
@@ -314,6 +317,33 @@ pub trait AsyncReadExt: AsyncRead {
         split::split(self)
     }
 
+    /// Creates an AsyncRead adapter which will read at most `limit` bytes
+    /// from the underlying reader.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(async_await)]
+    /// # futures::executor::block_on(async {
+    /// use futures::io::AsyncReadExt;
+    /// use std::io::Cursor;
+    ///
+    /// let reader = Cursor::new(&b"12345678"[..]);
+    /// let mut buffer = [0; 5];
+    ///
+    /// let mut take = reader.take(4);
+    /// let n = take.read(&mut buffer).await?;
+    ///
+    /// assert_eq!(n, 4);
+    /// assert_eq!(&buffer, b"1234\0");
+    /// # Ok::<(), Box<dyn std::error::Error>>(()) }).unwrap();
+    /// ```
+    fn take(self, limit: u64) -> Take<Self>
```

Comment:
Did you consider having this take `&mut self` instead of `self`-by-value?

Reply:
My (possibly wrong!) assumption is that `Take` is most often used as a "cap" and there's no interest in the rest of the underlying stream. I believe it would still be possible to write `(&mut reader).take(5)`, which would allow for the reuse of `reader` after dropping the `Take` instance. I can add a test for that if it's desired.

Reply:
Yep, that is possible-- both positions seem reasonable to me, so I'm happy to defer to your judgement here.

Reply:
This test passes with the current API:

```rust
let mut reader = Cursor::new(&b"1234567"[..]);
let mut buffer = [0; 5];
                                               
{
    let mut take = (&mut reader).take(4);
    let n = take.read(&mut buffer).await?;
                                               
    assert_eq!(n, 4);
    assert_eq!(&buffer, b"1234\0");
}
                                               
let n = reader.read(&mut buffer).await?;
assert_eq!(n, 3);
assert_eq!(&buffer, b"5674\0");
```

I did try out a `Take` that uses `&mut R` instead of `R`, but I quickly realized that it requires the caller of `take()` to keep the original reader around -- which is much less flexible for consumers.

Reply:
(Personally, I like this to have [the same signature (take `self`)](https://doc.rust-lang.org/nightly/std/io/trait.Read.html#method.take) as std.)